Priority

mikebyrneprth opened the pull request.

jessiagee reviewed (commented):
You'll want to remove all of the .DS_Store files, they aren't necessary

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      end
+
+      def purchase(amount, credit_card, options = {})
+        # byebug
```
Remove this comment

mikebyrneprth replied:
Removed

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        isSettleFunds = true
+        add_bank_amount_purchase(params, amount, authOnly)
+        add_credit_card(params, credit_card, action)
+        add_type_merchant_purchase(params, options[:merchant], isSettleFunds, options)
```
You can pass the `isSettleFunds` here as `true` if this is the only place that it's actually used in the method, same for `authOnly` above

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      end
+
+      def authorize(amount, credit_card, options = {})
+        #   byebug
```
Remove comment

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        isSettleFunds = false
+        add_bank_amount_purchase(params, amount, authOnly)
+        add_credit_card(params, credit_card, action)
+        add_type_merchant_purchase(params, options[:merchant], isSettleFunds, options)
```
You can pass the `isSettleFunds` here as true if this is the only place that it's actually used in the method, same for `authOnly` above

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      end
+
+      def refund(amount, credit_card, authCode, options, key_secret)
+        #   byebug
```
Remove comment

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      end
+
+      def capture(amount, authorization, authCode, options = {})
+        #   byebug
```
Remove comment

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        add_bank_amount_refund(params, amount, authCode)
+        add_credit_card(params, credit_card, action)
+        add_type_merchant_refund(params, options)
+        commit(action, params, '', '', key_secret)
```
You can set this `action` param to 'refund' and remove the cached variable

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        add_bank_amount_purchase(params, amount, authOnly)
+        add_credit_card(params, credit_card, action)
+        add_type_merchant_purchase(params, options[:merchant], isSettleFunds, options)
+        commit(action, params, '', '', options)
```
You can set this `action` param to 'purchase' and remove the cached variable

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        add_bank_amount_purchase(params, amount, authOnly)
+        add_credit_card(params, credit_card, action)
+        add_type_merchant_purchase(params, options[:merchant], isSettleFunds, options)
+        commit(action, params, '', '', options)
```
You can set this `action` param to 'purchase' and remove the cached variable

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        params = {}
+        action = 'capture'
+        add_capture(params, amount, authorization, authCode, options)
+        commit(action, params, '', '', options)
```
You can set this `action` param to 'capture' and remove the cached variable

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        commit(action, params, '', '', key_secret)
+      end
+
+      def capture(amount, authorization, authCode, options = {})
```
We expect there to be three headers for `capture` as detailed in the [Contributing](https://github.com/activemerchant/active_merchant/wiki/contributing#generating-the-gateway-skeleton) section. You can pass in the `authCode` in the `options` hash.

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      end
+
+      def add_capture(params, amount, authorization, authCode, options)
+        # byebug
```
Remove comment

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      end
+
+      def void(iid, options)
+        # byebug
```
Remove comment

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      def void(iid, options)
+        # byebug
+        action = 'void'
+        commit(action, '', iid, '', options)
```
You can set this `action` param to 'void' and remove the cached variable

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        params['tenderType'] = 'Card'
+      end
+
+      def void(iid, options)
```
Is this `iid` correct or should it be `id`? Also define it as `options = {}`

jessiagee replied:
@mikebyrneprth follow up?

jessiagee replied:
@mikebyrneprth still have a question about `iid` here ^

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        commit(action, params, '', '', options)
+      end
+
+      def add_capture(params, amount, authorization, authCode, options)
```
Is there a reason to do this separately outside of `capture`?

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      self.homepage_url = 'http://www.example.net/'
+      self.display_name = 'Priority'
+      # byebug
+      STANDARD_ERROR_CODE_MAPPING = {}
```
Remove this if empty and unused

jessiagee replied:
@mikebyrneprth follow up?

jessiagee replied:
@mikebyrneprth still have a question about STANDARD_ERROR_CODE_MAPPING ^

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      STANDARD_ERROR_CODE_MAPPING = {}
+
+      def initialize(options = {})
+        # requires!(options, :key, :secret)
```
Was this meant to be commented out?

jessiagee replied:
@mikebyrneprth still want to know if this was meant to be commented out ^

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        commit(action, params, '', '', options)
+      end
+
+      def refund(amount, credit_card, authCode, options, key_secret)
```
The method signature should match `refund(money, tx_reference, options = {})` as per the [Contributing](https://github.com/activemerchant/active_merchant/wiki/contributing#generating-the-gateway-skeleton) info - anything outside of the first two params can go into `options`

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+      def verify(creditcardnumber)
+        # byebug
+        #   test = request_verify_headers
```
Remove comments

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        commit(action, '', iid, '', options)
+      end
+
+      def verify(creditcardnumber)
```
update to `def verify(creditcard)`

jessiagee replied:
@mikebyrneprth follow up?

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        # byebug
+        #   test = request_verify_headers
+        action = 'verify'
+        commit(action, '', '', creditcardnumber, '')
```
You can set this `action` param to 'verify' and remove the cached variable

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        commit(action, '', '', creditcardnumber, '')
+      end
+
+      def getpaymentstatus(batchid, options)
```
Any method defs and params with spaces should use `_` - `def get_payment_status`

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+      def getpaymentstatus(batchid, options)
+        action = 'getpaymentstatus'
+        commit(action, batchid, '', '', options)
```
You can set this `action` param to 'getpaymentstatus' and remove the cached variable

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+      def closebatch(batchid, options)
+        action = 'closebatch'
+        commit(action, batchid, '', '', options)
```
You can set this `action` param to 'closebatch' and remove the cached variable

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        commit(action, batchid, '', '', options)
+      end
+
+      def closebatch(batchid, options)
```
Any method defs and params with spaces should use `_` - `def close_batch`

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      end
+      # private
+
+      def add_customer_data(post, options); end
```
Remove empty methods

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+      def add_customer_data(post, options); end
+
+      def add_address(post, creditcard); end
```
Remove empty methods

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      def add_bank_amount_refund(params, amount, authCode)
+        params['amount'] = amount
+        params['authCode'] = authCode
+        params['authMessage'] = 'Approved or completed successfully.'
```
Should this auth message be something we set? This looks like a message you would expect to get back from the server.

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      def add_credit_card(params, credit_card, action)
+        return unless credit_card
+
+        # byebug
```
Remove comment

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+        # byebug
+        params['cardAccount'] ||= {}
+        card_details = params['cardAccount'] = {}
```
Combine this with above line 

`card_details = params['cardAccount'] if params['cardAccount']`

jessiagee replied:
@mikebyrneprth follow up?

jessiagee replied:
@mikebyrneprth still looking for information here ^

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+        case action
+        when 'purchase'
+          card_details['avsStreet'] = '1'
```
Is the expectation the `avsStreet` and `avsZip` will always be these values on purchase?

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+          card_details['last4'] = credit_card['last4']
+          card_details['token'] = credit_card['token']
+        end
+        # byebug
```
Remove comment

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        authCode: nil,
+        authOnly: false,
+        bankAccount: nil,
+        # cardAccount: {
```
Remove comment

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        "#{format(credit_card.month, :two_digits)}/#{format(credit_card.year, :two_digits)}"
+      end
+
+      @request_params = {
```
Is this actually used somewhere?

jessiagee replied:
@mikebyrneprth follow up?

jessiagee replied:
@mikebyrneprth still need to know if this is used anywhere else ^

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        params['bankAccount'] = nil
+
+        params['purchases'] = purchases
+        # tax1 = {}
```
Remove commented out code

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        params['paymentToken'] = options['cardAccount']['token']
+
+        params['posData'] = options['posData']
+        # posdata = params['posData'] = {}
```
Remove commented out code

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+        post[:currency] = (options[:currency] || currency(money))
+      end
+
+      def add_payment(post, payment); end
```
Remove empty methods

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+              parse(ssl_invoke(action, params, iid, creditcardnumber, options))
+            end
+          rescue ResponseError => e
+            #   byebug
```
Remove comment

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+            #   byebug
+            parse(e.response.body)
+          end
+        #   byebug
```
Remove comment

mikebyrneprth replied:
Removed

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+      def handle_response(response)
+        #   byebug
+        # return response.body
```
Remove comment

mikebyrneprth replied:
Removed

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      end
+
+      def ssl_invoke(action, params, refnumber, creditcardnumber, options)
+        # byebug
```
Remove comments

mikebyrneprth replied:
Removed

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+      def ssl_invoke(action, params, refnumber, creditcardnumber, options)
+        # byebug
+        if action == 'void'
```
This could be converted to a switch statement for easier reading

mikebyrneprth replied:
Not done

jessiagee replied:
@mikebyrneprth thoughts on reason for not wanting to convert to a switch statement?

jessiagee replied:
@mikebyrneprth still looking for insight into this ^

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      end
+
+      def success_from(response)
+        # byebug
```
Remove comment

mikebyrneprth replied:
Done

mikebyrneprth commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+      class_attribute :test_url_verify, :live_url_verify, :test_auth, :live_auth, :test_env_verify, :live_env_verify, :test_url_batch, :live_url_batch, :merchant
+
+      self.test_url_verify = 'Contact priority for url'
```
Will email you.

jessiagee replied:
@mikebyrneprth have you sent that e-mail? Wanted to make sure I didn't miss it.

mikebyrneprth commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      self.test_env_verify = 'qa-aus'
+      self.live_env_verify = 'qa-aus'
+
+      require 'byebug'
```
Removed

mikebyrneprth commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+      self.default_currency = 'USD'
+      self.supported_cardtypes = %i[visa master american_express discover]
+
+      self.homepage_url = 'http://www.example.net/'
```
Updated as requested

mikebyrneprth commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,471 @@
+
+      self.homepage_url = 'http://www.example.net/'
+      self.display_name = 'Priority'
+      # byebug
```
Removed.

mikebyrneprth commented:
Most changes completed. Will email other details

jessiagee reviewed (changes requested):
> You'll want to remove all of the .DS_Store files, they aren't necessary

@mikebyrneprth just wanted to let you know I'm still seeing these in latest ^^

mikebyrneprth commented:
Jessi

I have pushed updated code.

Thanks

Mike.

On Mon, Sep 13, 2021 at 9:47 AM Jessi ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> You'll want to remove all of the .DS_Store files, they aren't necessary
>
> @mikebyrneprth <https://github.com/mikebyrneprth> just wanted to let you
> know I'm still seeing these in latest ^^
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406525>
> :
>
> > @@ -0,0 +1,471 @@
> +module ActiveMerchant #:nodoc:
> +  module Billing #:nodoc:
> +    class PriorityGateway < Gateway
> +      self.test_url = 'https://qa.api.mxmerchant.com/checkout/v3/payment'
> +      self.live_url = 'https://qa.api.mxmerchant.com/checkout/v3/payment'
> +
> +      class_attribute :test_url_verify, :live_url_verify, :test_auth, :live_auth, :test_env_verify, :live_env_verify, :test_url_batch, :live_url_batch, :merchant
> +
> +      self.test_url_verify = 'Contact priority for url'
>
> @mikebyrneprth <https://github.com/mikebyrneprth> have you sent that
> e-mail? Wanted to make sure I didn't miss it.
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406808>
> :
>
> > +        add_capture(params, amount, authorization, authCode, options)
> +        commit(action, params, '', '', options)
> +      end
> +
> +      def add_capture(params, amount, authorization, authCode, options)
> +        # byebug
> +        params['amount'] = amount
> +        params['authCode'] = authCode
> +        params['merchantId'] = options[:merchant]
> +        params['paymentToken'] = authorization
> +        params['shouldGetCreditCardLevel'] = true
> +        params['source'] = 'Spreedly'
> +        params['tenderType'] = 'Card'
> +      end
> +
> +      def void(iid, options)
>
> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406733>
> :
>
> > +
> +      self.test_auth = 'Contact priority for token'
> +      self.live_auth = 'Contact priority for token'
> +      self.test_env_verify = 'qa-aus'
> +      self.live_env_verify = 'qa-aus'
> +
> +      require 'byebug'
> +
> +      self.supported_countries = ['US']
> +      self.default_currency = 'USD'
> +      self.supported_cardtypes = %i[visa master american_express discover]
> +
> +      self.homepage_url = 'http://www.example.net/'
> +      self.display_name = 'Priority'
> +      # byebug
> +      STANDARD_ERROR_CODE_MAPPING = {}
>
> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406856>
> :
>
> > +        params['amount'] = amount
> +        params['authCode'] = authCode
> +        params['merchantId'] = options[:merchant]
> +        params['paymentToken'] = authorization
> +        params['shouldGetCreditCardLevel'] = true
> +        params['source'] = 'Spreedly'
> +        params['tenderType'] = 'Card'
> +      end
> +
> +      def void(iid, options)
> +        # byebug
> +        action = 'void'
> +        commit(action, '', iid, '', options)
> +      end
> +
> +      def verify(creditcardnumber)
>
> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406919>
> :
>
> > +      def add_bank_amount_refund(params, amount, authCode)
> +        params['amount'] = amount
> +        params['authCode'] = authCode
> +        params['authMessage'] = 'Approved or completed successfully.'
> +        params['authOnly'] = false
> +        params['availableAuthAmount'] = 0
> +        params['batch'] = nil
> +        params['batchId'] = nil
> +      end
> +
> +      def add_credit_card(params, credit_card, action)
> +        return unless credit_card
> +
> +        # byebug
> +        params['cardAccount'] ||= {}
> +        card_details = params['cardAccount'] = {}
>
> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707407000>
> :
>
> > +          card_details['expiryMonth'] = credit_card['expiryMonth']
> +          card_details['expiryYear'] = credit_card['expiryYear']
> +          card_details['hasContract'] = credit_card['hasContract']
> +          card_details['isCorp'] = credit_card['isCorp']
> +          card_details['isDebit'] = credit_card['isDebit']
> +          card_details['last4'] = credit_card['last4']
> +          card_details['token'] = credit_card['token']
> +        end
> +        # byebug
> +      end
> +
> +      def expdate(credit_card)
> +        "#{format(credit_card.month, :two_digits)}/#{format(credit_card.year, :two_digits)}"
> +      end
> +
> +      @request_params = {
>
> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707407392>
> :
>
> > +        # return response.body
> +        if response.code != '204' && (200...300).cover?(response.code.to_i)
> +          response.body
> +        else
> +
> +          if response.code == '204' || response == ''
> +            response.body = { 'code' => '204' }
> +          else
> +            raise ResponseError.new(response)
> +          end
> +        end
> +      end
> +
> +      def ssl_invoke(action, params, refnumber, creditcardnumber, options)
> +        # byebug
> +        if action == 'void'
>
> @mikebyrneprth <https://github.com/mikebyrneprth> thoughts on reason for
> not wanting to convert to a switch statement?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/activemerchant/active_merchant/pull/4080#pullrequestreview-752849267>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVE5OWIB7OOLJ2IHY5MKPH3UBYFGZANCNFSM5B4SDAUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

-- 


PROPRIETARY
/ CONFIDENTIALITY NOTICE





This message (including any 
attachments) contains information
that is proprietary and confidential to 
PRIORITY TECHNOLOGY HOLDINGS, INC. and its
affiliates and subsidiaries, 
including the sender hereof, and is for the sole
use of the intended 
recipients. If you are not an intended recipient, you may
not read, print, 
retain, use, copy, distribute, forward or disclose to anyone
this message 
or any information contained in this message (including any
attachments). 
If you have received this message in error, please advise the
sender of 
this error by reply e-mail, and please destroy all copies of this
message 
(including any attachments).

mikebyrneprth commented:
Jessie

Also remove .DS_store files.

Where you able to receive all instructions for setting up a MX Merchant
sandbox account?

MIke



On Wed, Sep 15, 2021 at 4:19 PM Mike Byrne ***@***.***> wrote:

> Jessi
>
> I have pushed updated code.
>
> Thanks
>
> Mike.
>
> On Mon, Sep 13, 2021 at 9:47 AM Jessi ***@***.***> wrote:
>
>> ***@***.**** requested changes on this pull request.
>>
>> You'll want to remove all of the .DS_Store files, they aren't necessary
>>
>> @mikebyrneprth <https://github.com/mikebyrneprth> just wanted to let you
>> know I'm still seeing these in latest ^^
>> ------------------------------
>>
>> In lib/active_merchant/billing/gateways/priority.rb
>> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406525>
>> :
>>
>> > @@ -0,0 +1,471 @@
>> +module ActiveMerchant #:nodoc:
>> +  module Billing #:nodoc:
>> +    class PriorityGateway < Gateway
>> +      self.test_url = 'https://qa.api.mxmerchant.com/checkout/v3/payment'
>> +      self.live_url = 'https://qa.api.mxmerchant.com/checkout/v3/payment'
>> +
>> +      class_attribute :test_url_verify, :live_url_verify, :test_auth, :live_auth, :test_env_verify, :live_env_verify, :test_url_batch, :live_url_batch, :merchant
>> +
>> +      self.test_url_verify = 'Contact priority for url'
>>
>> @mikebyrneprth <https://github.com/mikebyrneprth> have you sent that
>> e-mail? Wanted to make sure I didn't miss it.
>> ------------------------------
>>
>> In lib/active_merchant/billing/gateways/priority.rb
>> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406808>
>> :
>>
>> > +        add_capture(params, amount, authorization, authCode, options)
>> +        commit(action, params, '', '', options)
>> +      end
>> +
>> +      def add_capture(params, amount, authorization, authCode, options)
>> +        # byebug
>> +        params['amount'] = amount
>> +        params['authCode'] = authCode
>> +        params['merchantId'] = options[:merchant]
>> +        params['paymentToken'] = authorization
>> +        params['shouldGetCreditCardLevel'] = true
>> +        params['source'] = 'Spreedly'
>> +        params['tenderType'] = 'Card'
>> +      end
>> +
>> +      def void(iid, options)
>>
>> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
>> ------------------------------
>>
>> In lib/active_merchant/billing/gateways/priority.rb
>> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406733>
>> :
>>
>> > +
>> +      self.test_auth = 'Contact priority for token'
>> +      self.live_auth = 'Contact priority for token'
>> +      self.test_env_verify = 'qa-aus'
>> +      self.live_env_verify = 'qa-aus'
>> +
>> +      require 'byebug'
>> +
>> +      self.supported_countries = ['US']
>> +      self.default_currency = 'USD'
>> +      self.supported_cardtypes = %i[visa master american_express discover]
>> +
>> +      self.homepage_url = 'http://www.example.net/'
>> +      self.display_name = 'Priority'
>> +      # byebug
>> +      STANDARD_ERROR_CODE_MAPPING = {}
>>
>> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
>> ------------------------------
>>
>> In lib/active_merchant/billing/gateways/priority.rb
>> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406856>
>> :
>>
>> > +        params['amount'] = amount
>> +        params['authCode'] = authCode
>> +        params['merchantId'] = options[:merchant]
>> +        params['paymentToken'] = authorization
>> +        params['shouldGetCreditCardLevel'] = true
>> +        params['source'] = 'Spreedly'
>> +        params['tenderType'] = 'Card'
>> +      end
>> +
>> +      def void(iid, options)
>> +        # byebug
>> +        action = 'void'
>> +        commit(action, '', iid, '', options)
>> +      end
>> +
>> +      def verify(creditcardnumber)
>>
>> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
>> ------------------------------
>>
>> In lib/active_merchant/billing/gateways/priority.rb
>> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707406919>
>> :
>>
>> > +      def add_bank_amount_refund(params, amount, authCode)
>> +        params['amount'] = amount
>> +        params['authCode'] = authCode
>> +        params['authMessage'] = 'Approved or completed successfully.'
>> +        params['authOnly'] = false
>> +        params['availableAuthAmount'] = 0
>> +        params['batch'] = nil
>> +        params['batchId'] = nil
>> +      end
>> +
>> +      def add_credit_card(params, credit_card, action)
>> +        return unless credit_card
>> +
>> +        # byebug
>> +        params['cardAccount'] ||= {}
>> +        card_details = params['cardAccount'] = {}
>>
>> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
>> ------------------------------
>>
>> In lib/active_merchant/billing/gateways/priority.rb
>> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707407000>
>> :
>>
>> > +          card_details['expiryMonth'] = credit_card['expiryMonth']
>> +          card_details['expiryYear'] = credit_card['expiryYear']
>> +          card_details['hasContract'] = credit_card['hasContract']
>> +          card_details['isCorp'] = credit_card['isCorp']
>> +          card_details['isDebit'] = credit_card['isDebit']
>> +          card_details['last4'] = credit_card['last4']
>> +          card_details['token'] = credit_card['token']
>> +        end
>> +        # byebug
>> +      end
>> +
>> +      def expdate(credit_card)
>> +        "#{format(credit_card.month, :two_digits)}/#{format(credit_card.year, :two_digits)}"
>> +      end
>> +
>> +      @request_params = {
>>
>> @mikebyrneprth <https://github.com/mikebyrneprth> follow up?
>> ------------------------------
>>
>> In lib/active_merchant/billing/gateways/priority.rb
>> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r707407392>
>> :
>>
>> > +        # return response.body
>> +        if response.code != '204' && (200...300).cover?(response.code.to_i)
>> +          response.body
>> +        else
>> +
>> +          if response.code == '204' || response == ''
>> +            response.body = { 'code' => '204' }
>> +          else
>> +            raise ResponseError.new(response)
>> +          end
>> +        end
>> +      end
>> +
>> +      def ssl_invoke(action, params, refnumber, creditcardnumber, options)
>> +        # byebug
>> +        if action == 'void'
>>
>> @mikebyrneprth <https://github.com/mikebyrneprth> thoughts on reason for
>> not wanting to convert to a switch statement?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/activemerchant/active_merchant/pull/4080#pullrequestreview-752849267>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AVE5OWIB7OOLJ2IHY5MKPH3UBYFGZANCNFSM5B4SDAUQ>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>>
>

-- 


PROPRIETARY
/ CONFIDENTIALITY NOTICE





This message (including any 
attachments) contains information
that is proprietary and confidential to 
PRIORITY TECHNOLOGY HOLDINGS, INC. and its
affiliates and subsidiaries, 
including the sender hereof, and is for the sole
use of the intended 
recipients. If you are not an intended recipient, you may
not read, print, 
retain, use, copy, distribute, forward or disclose to anyone
this message 
or any information contained in this message (including any
attachments). 
If you have received this message in error, please advise the
sender of 
this error by reply e-mail, and please destroy all copies of this
message 
(including any attachments).

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,387 @@
+      end
+
+      def request_verify_headers(jwt)
+        auth = test? ? self.test_auth : self.live_auth
```
This `auth` variable doesn't appear to be used anywhere, is it necessary?

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,387 @@
+        commit('closebatch', batchid, '', '', options)
+      end
+
+      def createjwt(options)
```
Any method defs and params with spaces should use _ - `def create_jwt(options)`

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,387 @@
+        [{ taxRate: '0.0000', additionalTaxRate: nil, discountRate: nil }]
+      end
+
+      def add_type_merchant_purchase(params, merchant, isSettleFunds, options)
```
`isSettleFunds` should be `is_settle_funds` to match Ruby convention

jessiagee replied:
Hey @mikebyrneprth ! 

The `isSettleFunds` is now showing `is_Settle_Funds` instead of `is_settle_funds`. Also, it looks like any of the areas where they've been changed to Ruby convention of _'s was perhaps find and replaced, which potentially will cause what's sent to Priority to fail, for example:

`        params['is_Settle_Funds'] = is_Settle_Funds` <- I assume Priority would need this to be `params[isSettleFunds]`

jessiagee commented on test/fixtures.yml:
```diff
@@ -833,6 +833,11 @@ plugnpay:
 
+# Working credentials, no need to replace
+priority:
+  some_credential: SOMECREDENTIAL
```
Do these not need to be `login` and `password` or some other variation?

jessiagee commented on test/remote/gateways/remote_priority_test.rb:
```diff
@@ -0,0 +1,428 @@
+    # ruby -Itest test/unit/gateways/priority_test.rb -n test_successful_void
+
+    # Run specific remote test
+    # ruby -Itest test/remote/gateways/remote_priority_test.rb -n test_fail_refund_already_refunded_purchase_response
```
Thanks for these comments!! They're super helpful so I don't have to look it up 👏🏻 👏🏻

jessiagee commented on test/remote/gateways/remote_priority_test.rb:
```diff
@@ -0,0 +1,428 @@
+      billing_address: address,
+      key: 'Generated in MX Merchant for specific test merchant',
+      secret: 'Generated in MX Merchant for specific test merchant',
+      avsStreet: '666',
```
👿

jessiagee reviewed (changes requested):
Hi Mike,

I've added some more review comments here and also wanted to point out there are still a few from previous reviews that haven't been addressed/updated/etc.

I am going to follow back up on the ZenDesk ticket regarding the sandbox account.

Thanks!
Jessi

mikebyrneprth commented:
Hi Jessi

I have updated the code with regard to your comments.
Note will not use the "fixture.yaml" file for login credentials.
A new merchant will have to perform the following steps when they setup
with Priority gateway:
1). Login into MX Merchant.
[image: image.png]
2). Click on "Settngs" and select "Applications/ API" box.
[image: image.png]
3). Click the "Add Application" button and enter details. Click on the
"Generate Key" button.
[image: image.png]
4). Save the "key:" and "secret:" key.
5). Add key and secret key into "Spreedly" setup form.
So when the merchant performs a login on "Spreedly" the specific merchant
"key:" and "secret:" keys will loaded in.
These  "key:" and "secret:" keys will be used in the "authorization" header
whenever a call is made to the "Priority" end point.
[image: image.png]
See above highlighted lines shows how I currently test using the "key" and
"secret" keys. I suspect you will use a "initialization" function to load
these values.

Let me know if I can help with the setup of "MX Merchant" application in
our "Sandbox" environment or any other questions.

Regards

Mike

On Wed, Sep 22, 2021 at 4:37 PM Jessi ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Hi Mike,
>
> I've added some more review comments here and also wanted to point out
> there are still a few from previous reviews that haven't been
> addressed/updated/etc.
>
> I am going to follow back up on the ZenDesk ticket regarding the sandbox
> account.
>
> Thanks!
> Jessi
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r714323641>
> :
>
> > +        super
>
> +      end
>
> +
>
> +      def basic_auth(key, secret)
>
> +        Base64.strict_encode64("#{key}:#{secret}")
>
> +      end
>
> +
>
> +      def request_headers(options)
>
> +        {
>
> +          'Content-Type' => 'application/json',
>
> +          'Authorization' => "Basic #{basic_auth(options[:key], options[:secret])}"
>
> +        }
>
> +      end
>
> +
>
> +      def request_verify_headers(jwt)
>
> +        auth = test? ? self.test_auth : self.live_auth
>
>
> This auth variable doesn't appear to be used anywhere, is it necessary?
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r714324185>
> :
>
> > +        commit('verify', '', '', creditcardnumber, jwt)
>
> +      end
>
> +
>
> +      def supports_scrubbing?
>
> +        true
>
> +      end
>
> +
>
> +      def getpaymentstatus(batchid, options)
>
> +        commit('getpaymentstatus', batchid, '', '', options)
>
> +      end
>
> +
>
> +      def closebatch(batchid, options)
>
> +        commit('closebatch', batchid, '', '', options)
>
> +      end
>
> +
>
> +      def createjwt(options)
>
>
> Any method defs and params with spaces should use _ - def
> create_jwt(options)
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r714324916>
> :
>
> > +        mxAdvantageFeeLabel: '',
>
> +        paymentType: 'Sale',
>
> +        purchases: [{ taxRate: '0.0000', additionalTaxRate: nil, discountRate: nil }],
>
> +        shouldGetCreditCardLevel: true,
>
> +        shouldVaultCard: true,
>
> +        source: 'QuickPay',
>
> +        sourceZip: '94102',
>
> +        taxExempt: false,
>
> +        tenderType: 'Card',
>
> +        terminals: []
>
> +      }
>
> +      def purchases
>
> +        [{ taxRate: '0.0000', additionalTaxRate: nil, discountRate: nil }]
>
> +      end
>
> +
>
> +      def add_type_merchant_purchase(params, merchant, isSettleFunds, options)
>
>
> isSettleFunds should be is_settle_funds to match Ruby convention
> ------------------------------
>
> In test/fixtures.yml
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r714325383>
> :
>
> > @@ -833,6 +833,11 @@ plugnpay:
>
>    login: LOGIN
>
>    password: PASSWORD
>
>
>
> +# Working credentials, no need to replace
>
> +priority:
>
> +  some_credential: SOMECREDENTIAL
>
>
> Do these not need to be login and password or some other variation?
> ------------------------------
>
> In test/remote/gateways/remote_priority_test.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r714325832>
> :
>
> > @@ -0,0 +1,428 @@
>
> +require 'test_helper'
>
> +
>
> +class RemotePriorityTest < Test::Unit::TestCase
>
> +  def setup
>
> +    # Consumer API Key: Generated in MX Merchant for specific test merchant
>
> +    # Consumer API Secret:= Generated in MX Merchant for specific test merchant
>
> +
>
> +    # run command below to run tests in debug (byebug)
>
> +    # byebug -Itest test/unit/gateways/card_stream_test.rb
>
> +    #
>
> +    # bundle exec rake test:remote TEST=test/remote/gateways/remote_priority_test.rb
>
> +    # ruby -Itest test/unit/gateways/priority_test.rb -n test_successful_void
>
> +
>
> +    # Run specific remote test
>
> +    # ruby -Itest test/remote/gateways/remote_priority_test.rb -n test_fail_refund_already_refunded_purchase_response
>
>
> Thanks for these comments!! They're super helpful so I don't have to look
> it up 👏🏻 👏🏻
> ------------------------------
>
> In test/remote/gateways/remote_priority_test.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r714325987>
> :
>
> > +    # ruby -Itest test/unit/gateways/priority_test.rb -n test_successful_void
>
> +
>
> +    # Run specific remote test
>
> +    # ruby -Itest test/remote/gateways/remote_priority_test.rb -n test_fail_refund_already_refunded_purchase_response
>
> +    @gateway = PriorityGateway.new(fixtures(:priority))
>
> +
>
> +    # purchase params success
>
> +    @amount_purchase = 2.11
>
> +    @credit_card_purchase_success = credit_card('4111111111111111', month: '01', year: '2029', first_name: 'Marcus', last_name: 'Rashford', verification_value: '123')
>
> +
>
> +    @option_spr = {
>
> +      merchant: 514391592,
>
> +      billing_address: address,
>
> +      key: 'Generated in MX Merchant for specific test merchant',
>
> +      secret: 'Generated in MX Merchant for specific test merchant',
>
> +      avsStreet: '666',
>
>
> 👿
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/activemerchant/active_merchant/pull/4080#pullrequestreview-761423581>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVE5OWOVUEQOEW5Q6DP5KITUDJEA3ANCNFSM5B4SDAUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

-- 


PROPRIETARY
/ CONFIDENTIALITY NOTICE





This message (including any 
attachments) contains information
that is proprietary and confidential to 
PRIORITY TECHNOLOGY HOLDINGS, INC. and its
affiliates and subsidiaries, 
including the sender hereof, and is for the sole
use of the intended 
recipients. If you are not an intended recipient, you may
not read, print, 
retain, use, copy, distribute, forward or disclose to anyone
this message 
or any information contained in this message (including any
attachments). 
If you have received this message in error, please advise the
sender of 
this error by reply e-mail, and please destroy all copies of this
message 
(including any attachments).

jessiagee commented:
> Hi Jessi I have updated the code with regard to your comments. Note will not use the "fixture.yaml" file for login credentials. A new merchant will have to perform the following steps when they setup with Priority gateway: 1). Login into MX Merchant. [image: image.png] 2). Click on "Settngs" and select "Applications/ API" box. [image: image.png] 3). Click the "Add Application" button and enter details. Click on the "Generate Key" button. [image: image.png] 4). Save the "key:" and "secret:" key. 5). Add key and secret key into "Spreedly" setup form. So when the merchant performs a login on "Spreedly" the specific merchant "key:" and "secret:" keys will loaded in. These "key:" and "secret:" keys will be used in the "authorization" header whenever a call is made to the "Priority" end point. [image: image.png] See above highlighted lines shows how I currently test using the "key" and "secret" keys. I suspect you will use a "initialization" function to load these values. Let me know if I can help with the setup of "MX Merchant" application in our "Sandbox" environment or any other questions. Regards Mike

Hey @mikebyrneprth thanks for the info here re: merchant sandbox testing. I haven't gotten a new link yet so once I'll do I'll work on testing remote tests. 

I've left a comment on the code above and also wanted to see what your thoughts are re: the other comments that are still open.  Thanks!

Jessi

jessiagee reviewed (changes requested):
Comments already left

jessiagee commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,386 @@
+        params['cardPresent'] = false
+        params['cardPresentType'] = 'CardNotPresent'
+        params['isAuth'] = true
+        params['is_settle_funds'] = is_settle_funds
```
@mikebyrneprth is PPS expecting `is_settle_funds` or are they expecting `isSettleFunds`? If the latter, please update the areas in the code that will go to PPS in the `def commit` but leave anything in Active Merchant from Spreedly as `is_settle_funds` for Ruby convention

jessiagee reviewed (changes requested):
Looks like there's seven comments left to address and then we should be good to test for Spreedly's side 🎉

mikebyrneprth commented:
Jessi .

See responses in red.

In lib/active_merchant/billing/gateways/priority.rb
<https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718656052>
:

> +        shouldVaultCard: true,

+        source: 'QuickPay',

+        sourceZip: '94102',

+        taxExempt: false,

+        tenderType: 'Card',

+        terminals: []

+      }

Removed above.

+      def purchases

+        [{ taxRate: '0.0000', additionalTaxRate: nil, discountRate: nil }]

+      end

+

+      def add_type_merchant_purchase(params, merchant,
is_settle_funds, options)

+        params['cardPresent'] = false

+        params['cardPresentType'] = 'CardNotPresent'

+        params['isAuth'] = true

+        params['is_settle_funds'] = is_settle_funds


@mikebyrneprth <https://github.com/mikebyrneprth> is PPS expecting
is_settle_funds or are they expecting isSettleFunds? If the latter, please
update the areas in the code that will go to PPS in the def commit but
leave anything in Active Merchant from Spreedly as is_settle_funds for Ruby
convention

Updated input parameters to isSettleFunds for parameters expected by
Priority. Input var is is_settle_funds
------------------------------

In lib/active_merchant/billing/gateways/priority.rb
<https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718653274>
:

> +        add_capture(params, amount, authorization, authCode, options)

+        commit(action, params, '', '', options)

+      end

+

+      def add_capture(params, amount, authorization, authCode, options)

+        # byebug

+        params['amount'] = amount

+        params['authCode'] = authCode

+        params['merchantId'] = options[:merchant]

+        params['paymentToken'] = authorization

+        params['shouldGetCreditCardLevel'] = true

+        params['source'] = 'Spreedly'

+        params['tenderType'] = 'Card'

+      end

+

+      def void(iid, options)


@mikebyrneprth <https://github.com/mikebyrneprth> still have a question
about iid here ^

"iid" is valid field name in Priority. Not a type error!
------------------------------

In lib/active_merchant/billing/gateways/priority.rb
<https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718653481>
:

> +

+      self.test_auth = 'Contact priority for token'

+      self.live_auth = 'Contact priority for token'

+      self.test_env_verify = 'qa-aus'

+      self.live_env_verify = 'qa-aus'

+

+      require 'byebug'

+

+      self.supported_countries = ['US']

+      self.default_currency = 'USD'

+      self.supported_cardtypes = %i[visa master american_express discover]

+

+      self.homepage_url = 'http://www.example.net/'

+      self.display_name = 'Priority'

+      # byebug

+      STANDARD_ERROR_CODE_MAPPING = {}


@mikebyrneprth <https://github.com/mikebyrneprth> still have a question
about STANDARD_ERROR_CODE_MAPPING ^

Removed STANDARD_ERROR_CODE_MAPPING variable.
------------------------------

In lib/active_merchant/billing/gateways/priority.rb
<https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718653778>
:

> +      self.test_env_verify = 'qa-aus'

+      self.live_env_verify = 'qa-aus'

+

+      require 'byebug'

+

+      self.supported_countries = ['US']

+      self.default_currency = 'USD'

+      self.supported_cardtypes = %i[visa master american_express discover]

+

+      self.homepage_url = 'http://www.example.net/'

+      self.display_name = 'Priority'

+      # byebug

+      STANDARD_ERROR_CODE_MAPPING = {}

+

+      def initialize(options = {})

+        # requires!(options, :key, :secret)


@mikebyrneprth <https://github.com/mikebyrneprth> still want to know if
this was meant to be commented out ^

Left this is as a marker. Each merchant when they logon to Spreedly will
have a unique API key. This unique key for each merchant is generated in MX
Merchant. See my previous email about this. Spreedly will hold these keys
on your database. Each time a merchant logons to Spreedly they will load
the unique keys specific for that merchant. The "key" and "secret" will be
populated via "Spreedly" merchant logon process.
[image: image.png]


------------------------------

In lib/active_merchant/billing/gateways/priority.rb
<https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718654052>
:

> +      def add_bank_amount_refund(params, amount, authCode)

+        params['amount'] = amount

+        params['authCode'] = authCode

+        params['authMessage'] = 'Approved or completed successfully.'

+        params['authOnly'] = false

+        params['availableAuthAmount'] = 0

+        params['batch'] = nil

+        params['batchId'] = nil

+      end

+

+      def add_credit_card(params, credit_card, action)

+        return unless credit_card

+

+        # byebug

+        params['cardAccount'] ||= {}

+        card_details = params['cardAccount'] = {}


@mikebyrneprth <https://github.com/mikebyrneprth> still looking for
information here ^

I tried your suggested code. It caused issues.
------------------------------

In lib/active_merchant/billing/gateways/priority.rb
<https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718654375>
:

> +          card_details['expiryMonth'] = credit_card['expiryMonth']

+          card_details['expiryYear'] = credit_card['expiryYear']

+          card_details['hasContract'] = credit_card['hasContract']

+          card_details['isCorp'] = credit_card['isCorp']

+          card_details['isDebit'] = credit_card['isDebit']

+          card_details['last4'] = credit_card['last4']

+          card_details['token'] = credit_card['token']

+        end

+        # byebug

+      end

+

+      def expdate(credit_card)

+        "#{format(credit_card.month,
:two_digits)}/#{format(credit_card.year, :two_digits)}"

+      end

+

+      @request_params = {


@mikebyrneprth <https://github.com/mikebyrneprth> still need to know if
this is used anywhere else ^

I have removed ***@***.***_params"
------------------------------

In lib/active_merchant/billing/gateways/priority.rb
<https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718654632>
:

> +        # return response.body

+        if response.code != '204' && (200...300).cover?(response.code.to_i)

+          response.body

+        else

+

+          if response.code == '204' || response == ''

+            response.body = { 'code' => '204' }

+          else

+            raise ResponseError.new(response)

+          end

+        end

+      end

+

+      def ssl_invoke(action, params, refnumber, creditcardnumber, options)

+        # byebug

+        if action == 'void'


@mikebyrneprth <https://github.com/mikebyrneprth> still looking for insight
into this ^
[image: image.png]

Variable used to pass to "verify" end point.

I have removed all  references to "byebug" in my pull request.


[image: image.png]

I have escalated the request for access to our Sandbox applications (MXC
and MX Merchant).

Mike



On Wed, Sep 29, 2021 at 10:48 AM Jessi ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Looks like there's seven comments left to address and then we should be
> good to test for Spreedly's side 🎉
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718656052>
> :
>
> > +        shouldVaultCard: true,
>
> +        source: 'QuickPay',
>
> +        sourceZip: '94102',
>
> +        taxExempt: false,
>
> +        tenderType: 'Card',
>
> +        terminals: []
>
> +      }
>
> +      def purchases
>
> +        [{ taxRate: '0.0000', additionalTaxRate: nil, discountRate: nil }]
>
> +      end
>
> +
>
> +      def add_type_merchant_purchase(params, merchant, is_settle_funds, options)
>
> +        params['cardPresent'] = false
>
> +        params['cardPresentType'] = 'CardNotPresent'
>
> +        params['isAuth'] = true
>
> +        params['is_settle_funds'] = is_settle_funds
>
>
> @mikebyrneprth <https://github.com/mikebyrneprth> is PPS expecting
> is_settle_funds or are they expecting isSettleFunds? If the latter,
> please update the areas in the code that will go to PPS in the def commit
> but leave anything in Active Merchant from Spreedly as is_settle_funds
> for Ruby convention
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718653274>
> :
>
> > +        add_capture(params, amount, authorization, authCode, options)
>
> +        commit(action, params, '', '', options)
>
> +      end
>
> +
>
> +      def add_capture(params, amount, authorization, authCode, options)
>
> +        # byebug
>
> +        params['amount'] = amount
>
> +        params['authCode'] = authCode
>
> +        params['merchantId'] = options[:merchant]
>
> +        params['paymentToken'] = authorization
>
> +        params['shouldGetCreditCardLevel'] = true
>
> +        params['source'] = 'Spreedly'
>
> +        params['tenderType'] = 'Card'
>
> +      end
>
> +
>
> +      def void(iid, options)
>
>
> @mikebyrneprth <https://github.com/mikebyrneprth> still have a question
> about iid here ^
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718653481>
> :
>
> > +
>
> +      self.test_auth = 'Contact priority for token'
>
> +      self.live_auth = 'Contact priority for token'
>
> +      self.test_env_verify = 'qa-aus'
>
> +      self.live_env_verify = 'qa-aus'
>
> +
>
> +      require 'byebug'
>
> +
>
> +      self.supported_countries = ['US']
>
> +      self.default_currency = 'USD'
>
> +      self.supported_cardtypes = %i[visa master american_express discover]
>
> +
>
> +      self.homepage_url = 'http://www.example.net/'
>
> +      self.display_name = 'Priority'
>
> +      # byebug
>
> +      STANDARD_ERROR_CODE_MAPPING = {}
>
>
> @mikebyrneprth <https://github.com/mikebyrneprth> still have a question
> about STANDARD_ERROR_CODE_MAPPING ^
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718653778>
> :
>
> > +      self.test_env_verify = 'qa-aus'
>
> +      self.live_env_verify = 'qa-aus'
>
> +
>
> +      require 'byebug'
>
> +
>
> +      self.supported_countries = ['US']
>
> +      self.default_currency = 'USD'
>
> +      self.supported_cardtypes = %i[visa master american_express discover]
>
> +
>
> +      self.homepage_url = 'http://www.example.net/'
>
> +      self.display_name = 'Priority'
>
> +      # byebug
>
> +      STANDARD_ERROR_CODE_MAPPING = {}
>
> +
>
> +      def initialize(options = {})
>
> +        # requires!(options, :key, :secret)
>
>
> @mikebyrneprth <https://github.com/mikebyrneprth> still want to know if
> this was meant to be commented out ^
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718654052>
> :
>
> > +      def add_bank_amount_refund(params, amount, authCode)
>
> +        params['amount'] = amount
>
> +        params['authCode'] = authCode
>
> +        params['authMessage'] = 'Approved or completed successfully.'
>
> +        params['authOnly'] = false
>
> +        params['availableAuthAmount'] = 0
>
> +        params['batch'] = nil
>
> +        params['batchId'] = nil
>
> +      end
>
> +
>
> +      def add_credit_card(params, credit_card, action)
>
> +        return unless credit_card
>
> +
>
> +        # byebug
>
> +        params['cardAccount'] ||= {}
>
> +        card_details = params['cardAccount'] = {}
>
>
> @mikebyrneprth <https://github.com/mikebyrneprth> still looking for
> information here ^
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718654375>
> :
>
> > +          card_details['expiryMonth'] = credit_card['expiryMonth']
>
> +          card_details['expiryYear'] = credit_card['expiryYear']
>
> +          card_details['hasContract'] = credit_card['hasContract']
>
> +          card_details['isCorp'] = credit_card['isCorp']
>
> +          card_details['isDebit'] = credit_card['isDebit']
>
> +          card_details['last4'] = credit_card['last4']
>
> +          card_details['token'] = credit_card['token']
>
> +        end
>
> +        # byebug
>
> +      end
>
> +
>
> +      def expdate(credit_card)
>
> +        "#{format(credit_card.month, :two_digits)}/#{format(credit_card.year, :two_digits)}"
>
> +      end
>
> +
>
> +      @request_params = {
>
>
> @mikebyrneprth <https://github.com/mikebyrneprth> still need to know if
> this is used anywhere else ^
> ------------------------------
>
> In lib/active_merchant/billing/gateways/priority.rb
> <https://github.com/activemerchant/active_merchant/pull/4080#discussion_r718654632>
> :
>
> > +        # return response.body
>
> +        if response.code != '204' && (200...300).cover?(response.code.to_i)
>
> +          response.body
>
> +        else
>
> +
>
> +          if response.code == '204' || response == ''
>
> +            response.body = { 'code' => '204' }
>
> +          else
>
> +            raise ResponseError.new(response)
>
> +          end
>
> +        end
>
> +      end
>
> +
>
> +      def ssl_invoke(action, params, refnumber, creditcardnumber, options)
>
> +        # byebug
>
> +        if action == 'void'
>
>
> @mikebyrneprth <https://github.com/mikebyrneprth> still looking for
> insight into this ^
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/activemerchant/active_merchant/pull/4080#pullrequestreview-766880503>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVE5OWKEKAG57WNTWMUZXKTUEMYODANCNFSM5B4SDAUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

-- 


PROPRIETARY
/ CONFIDENTIALITY NOTICE





This message (including any 
attachments) contains information
that is proprietary and confidential to 
PRIORITY TECHNOLOGY HOLDINGS, INC. and its
affiliates and subsidiaries, 
including the sender hereof, and is for the sole
use of the intended 
recipients. If you are not an intended recipient, you may
not read, print, 
retain, use, copy, distribute, forward or disclose to anyone
this message 
or any information contained in this message (including any
attachments). 
If you have received this message in error, please advise the
sender of 
this error by reply e-mail, and please destroy all copies of this
message 
(including any attachments).

jessiagee reviewed (changes requested):
Hey Mike! 

I've created some commits with updates to the setup that you can find [here](https://github.com/activemerchant/active_merchant/commit/cba6d46798f73236e112117fd4341b0f537694b8) and [here](https://github.com/activemerchant/active_merchant/commit/cba6d46798f73236e112117fd4341b0f537694b8). I've also updated the [the `fixtures.yml` file](https://github.com/activemerchant/active_merchant/commit/9a96c5f7d699cd6d1171d6baebadfec24159f27f).

While looking over your tests, I noticed that there's a lot of passing the key and secret in the options hash which is not necessary. In my commits, you can see that I fix up the `initialize` param in the `priority.rb` class so that it's throwing an error if `merchant_id`, `key`, or `secret` don't exist. The docs show `merchant_id` as a requirement throughout, so it's good to ensure that it exists in the initialization.

With this set up, a merchant will create a gateway like in [the Adyen set up](https://docs.spreedly.com/payment-gateways/adyen/#adding-an-adyen-gateway), passing in their MX Merchant key, secret, and merchant id, and then the `super` call will add that information to the gateway `@options` instance variable so that you won't have to keep passing it around in the options hash.

I updated most of the issues I found in the remote tests in my commits, but there are a few that will need to be addressed before we can approve the PR:

`test_fail_Authorize_threshold_exceeded`
`test_fail_purchase_threshold_exceeded`
`test_successful_refund_purchase_response`

You will also want to go through the unit tests and make sure that they work as expected. I believe a few of them require the user to go into the tests and make changes to them in order to test; i.e., if you want to test the `get_payment_status` you have to provide an actual batch_id. If there's no easy way to test a method with a unit test because you have to manually provide data, than keep that as a remote test and remove it from the unit tests. 

Anyone who pulls down the tests should be able to run them after updating their own copy of [the `fixtures.yml` file](https://github.com/activemerchant/active_merchant/commit/9a96c5f7d699cd6d1171d6baebadfec24159f27f) (assuming no whitelist IP needs). Due to the fact that MXMerchant does not provide public, static testing credentials, each individual will have to add their own credentials to that file to directly run Active Merchant tests but should not have to provide them otherwise.

Please let me know if you have any other questions!
Jessi

naashton commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,364 @@
+        params['settledDate'] = options['created']
+        params['shipToCountry'] = options['shipToCountry']
+        params['shouldGetCreditCardLevel'] = options['shouldGetCreditCardLevel']
+        params['source'] = 'Spreedly'
```
Because ActiveMerchant is a collaborative open-source project, this value should come from the `options` hash.

naashton commented on lib/active_merchant/billing/gateways/priority.rb:
```diff
@@ -0,0 +1,364 @@
+        end
+      end
+
+      def url(action, params, refnumber = '', creditcardnumber = '')
```
This method is more traditionally used to determine whether or not to use a production, test, or other hostname for the given gateway.

Routes can and probably should be handled in the `commit` method where the `action` can be used to determine the `uri` and any values that need to be appended to that route.
